# Food Delivery Prediction 

In [ ]:
I love to order my food online for home and work.(not that I have any option as well)
So, I want to predict whether my order will be on time or late beforehand.
I logged my previous 45 orders like this:
["bad_weather", "is_rush_hour", "km_distance", "urban_address", "late"]

# KNN Example

Let's write our data into a Panda's data frame.

In [4]:
import pandas as pd
import numpy as np

In [7]:
data = np.array([[0, 1, 5, 1, 0],
                [1, 0, 7, 0, 1],
                [0, 1, 2, 1, 0],
                [1, 1, 4.2, 1, 0],
                [0, 0, 7.8, 0, 1],
                [1, 0, 3.9, 1, 0],
                [0, 1, 4, 1, 0],
                [1, 1, 2, 0, 0],
                [0, 0, 3.5, 0, 1],
                [1, 0, 2.6, 1, 0],
                [0, 0, 4.1, 0, 1],
                [0, 1, 1.5, 0, 1],
                [1, 1, 1.75, 1, 0],
                [1, 0, 1.3, 0, 0],
                [1, 1, 2.1, 0, 0],
                [1, 1, 0.2, 1, 0],
                [1, 1, 5.2, 0, 1],
                [0, 1, 2, 1, 0],
                [1, 0, 5.5, 0, 1],
                [0, 0, 2, 1, 0],
                [1, 1, 1.7, 0, 0],
                [0, 1, 3, 1, 1],
                [1, 1, 1.9, 1, 0],
                [0, 1, 3.1, 0, 1],
                [0, 1, 2.3, 0, 0],
                [0, 0, 1.1, 1, 0],
                [1, 1, 2.5, 1, 1],
                [1, 1, 5, 0, 1],
                [1, 0, 7.5, 1, 1],
                [0, 0, 0.5, 1, 0],
                [0, 0, 1.5, 1, 0],
                [1, 0, 3.2, 1, 0],
                [0, 0, 2.15, 1, 0],
                [1, 1, 4.2, 0, 1],
                [1, 0, 6.5, 0, 1],
                [1, 0, 0.5, 0, 0],
                [0, 0, 3.5, 0, 1],
                [0, 0, 1.75, 0, 0],
                [1, 1, 5, 0, 1],
                [0, 0, 2, 1, 0],
                [0, 1, 1.3, 1, 1],
                [0, 1, 0.2, 0, 0],
                [1, 1, 2.2, 0, 0],
                [0, 1, 1.2, 1, 0],
                [1, 1, 4.2, 0, 1]])

# Creating the data frame with this data and the columns
delivery_data = pd.DataFrame(data, columns=["bad_weather", "is_rush_hour", "km_distance", "urban_address", "late"])

In [8]:
delivery_data

bad_weather  is_rush_hour  km_distance  urban_address  late
0           0.0           1.0         5.00            1.0   0.0
1           1.0           0.0         7.00            0.0   1.0
2           0.0           1.0         2.00            1.0   0.0
3           1.0           1.0         4.20            1.0   0.0
4           0.0           0.0         7.80            0.0   1.0
5           1.0           0.0         3.90            1.0   0.0
6           0.0           1.0         4.00            1.0   0.0
7           1.0           1.0         2.00            0.0   0.0
8           0.0           0.0         3.50            0.0   1.0
9           1.0           0.0         2.60            1.0   0.0
10          0.0           0.0         4.10            0.0   1.0
11          0.0           1.0         1.50            0.0   1.0
12          1.0           1.0         1.75            1.0   0.0
13          1.0           0.0         1.30            0.0   0.0
14          1.0           1.0         2.10            0.0   0.0
15          1.0           1.0         0.20            1.0   0.0
16          1.0           1.0         5.20            0.0   1.0
17          0.0           1.0         2.00            1.0   0.0
18          1.0           0.0         5.50            0.0   1.0
19          0.0           0.0         2.00            1.0   0.0
20          1.0           1.0         1.70            0.0   0.0
21          0.0           1.0         3.00            1.0   1.0
22          1.0           1.0         1.90            1.0   0.0
23          0.0           1.0         3.10            0.0   1.0
24          0.0           1.0         2.30            0.0   0.0
25          0.0           0.0         1.10            1.0   0.0
26          1.0           1.0         2.50            1.0   1.0
27          1.0           1.0         5.00            0.0   1.0
28          1.0           0.0         7.50            1.0   1.0
29          0.0           0.0         0.50            1.0   0.0
30          0.0           0.0         1.50            1.0   0.0
31          1.0           0.0         3.20            1.0   0.0
32          0.0           0.0         2.15            1.0   0.0
33          1.0           1.0         4.20            0.0   1.0
34          1.0           0.0         6.50            0.0   1.0
35          1.0           0.0         0.50            0.0   0.0
36          0.0           0.0         3.50            0.0   1.0
37          0.0           0.0         1.75            0.0   0.0
38          1.0           1.0         5.00            0.0   1.0
39          0.0           0.0         2.00            1.0   0.0
40          0.0           1.0         1.30            1.0   1.0
41          0.0           1.0         0.20            0.0   0.0
42          1.0           1.0         2.20            0.0   0.0
43          0.0           1.0         1.20            1.0   0.0
44          1.0           1.0         4.20            0.0   1.0

In [9]:
# This will print the first 5 rows
delivery_data.head()

bad_weather  is_rush_hour  km_distance  urban_address  late
0          0.0           1.0          5.0            1.0   0.0
1          1.0           0.0          7.0            0.0   1.0
2          0.0           1.0          2.0            1.0   0.0
3          1.0           1.0          4.2            1.0   0.0
4          0.0           0.0          7.8            0.0   1.0

Let's get the input and output of our problem.

In [10]:
target = delivery_data["late"].tolist()
input_data = delivery_data[["bad_weather", "is_rush_hour", "km_distance", "urban_address"]].values

__The KNN Model:__ Let's fit the model to our data.

We will use the Sklearn library's KNN function here.

In [14]:
from sklearn.neighbors import KNeighborsClassifier

# Below, we are using n_neighbors = 2. 
# This means the KNN will consider two other "closest" records to make a decision.
knn_classifier = KNeighborsClassifier(n_neighbors = 2)

# We will fit the model to our data
knn_classifier.fit(input_data, target)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=2, p=2,
                     weights='uniform')

Let's make some prediction with our model. Assume we have the following data:

In [16]:
import numpy as np

data = np.array([[0, 0, 2.1, 1]]) # bad_weather->0, is_rush_hour->0, km_distance->2.1 and urban_address->1

print(knn_classifier.predict(data))

[0.]


We predicted this delivery to be on time.

We can predict multiple records below.

In [19]:
import numpy as np

data = np.array([[0, 0, 2.1, 1], # bad_weather->0, is_rush_hour->0, km_distance->2.1 and urban_address->1
                 [0, 1, 5, 0],   # bad_weather->0, is_rush_hour->1, km_distance->5.0 and urban_address->0
                 [1, 1, 3.1, 1]  # bad_weather->1, is_rush_hour->1, km_distance->3.1 and urban_address->1
                ])

print(knn_classifier.predict(data))

[0. 0. 1.]


The last delivery is predicted to be late. The first two will be on time (Hopefully :D)

# Overfitting
In the above lines of code, we quickly saw how to use a simple ML model to make predictions about the outcome (Late or on time delivery) of a problem. In this example, we will focus on building Training, Validation and Test subsets and discuss their importance.

### Approach 1: KNN trained on the whole dataset
As a first model, we will fit a KNN model to our whole data. 

In [20]:
delivery_data.shape

(45, 5)

In [22]:
target = delivery_data["late"].tolist()
input_data = delivery_data[["bad_weather", "is_rush_hour", "km_distance", "urban_address"]].values

In [23]:
from sklearn.neighbors import KNeighborsClassifier

# Below, we are using n_neighbors = 1
# This means the KNN will consider one other "closest" record to make a decision.
knn_classifier = KNeighborsClassifier(n_neighbors = 1)

# We will fit the model to our data
knn_classifier.fit(input_data, target)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=1, p=2,
                     weights='uniform')

Let's make some prediction with our model. Assume we have the following data:

In [24]:
import numpy as np

data = np.array([[0, 0, 2.1, 1]]) # bad_weather->0, is_rush_hour->0, mile_distance->2.1 and urban_address->1

print(knn_classifier.predict(data))

[0.]


__How do we know whether this was a good or bad prediction?__ <br/>
If we don't have the correct label for this input, we won't know. Similarly, we won't have any idea about how good this model is. 

One thing we can do is to test this model with the data we used to train it. We will introduce the __accuracy__ metric here that measures how correctly it predicts the late or not late outcome.

In [25]:
from sklearn.metrics import accuracy_score

preds = knn_classifier.predict(input_data)

accuracy_score(target, preds)

1.0

We predicted the outcome with 100% accuracy.

### Approach 2: Using a test dataset
My model worked with 100% accuracy on the whole dataset. This might seem promising, but this doesn't tell us anything about our future orders. In order to see whether this model works with some new "unseen" orders, we need to reserve some data for test purposes. 

__Let's split our data into two sets: Training (85%) and Test (15%)__

This will give us 38 training records and 7 test records (total 45 records)

In [27]:
# Let's split our data into two sets: Training (85%) and Test (15%)
# This will give us 38 training records and 7 test records (total 45 records)

training_data = delivery_data.iloc[:38, :] # First 38
test_data = delivery_data.iloc[38:, :] # Remaining

# This will print the first 5 rows
training_data.head()

bad_weather  is_rush_hour  km_distance  urban_address  late
0          0.0           1.0          5.0            1.0   0.0
1          1.0           0.0          7.0            0.0   1.0
2          0.0           1.0          2.0            1.0   0.0
3          1.0           1.0          4.2            1.0   0.0
4          0.0           0.0          7.8            0.0   1.0

Fitting the KNN on training portion this time.

In [29]:
from sklearn.neighbors import KNeighborsClassifier

target = training_data["late"].tolist()
input_data = training_data[["bad_weather", "is_rush_hour", "km_distance", "urban_address"]].values

# Below, we are using n_neighbors = 1
# This means the KNN will consider two other "closest" records to make a decision.
knn_classifier = KNeighborsClassifier(n_neighbors = 1)

# We will fit the model to our data
knn_classifier.fit(input_data, target)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=1, p=2,
                     weights='uniform')

Let's see the accuracy on training data

In [31]:
from sklearn.metrics import accuracy_score

preds = knn_classifier.predict(input_data)

accuracy_score(target, preds)

1.0

Let's see the accuracy on test data

In [33]:
from sklearn.metrics import accuracy_score

test_target = test_data["late"].tolist()
test_input_data = test_data[["bad_weather", "is_rush_hour", "km_distance", "urban_address"]].values

test_preds = knn_classifier.predict(test_input_data)

accuracy_score(test_target, test_preds)

0.7142857142857143

### This doesn't look good! 
We only achieved 71% accuracy with our data that the model hasn't seen before. <br/>
Can we trust this model? Probably not.

### Let's explain what happened here. 
We experienced a common problem called __"Overfitting"__. This means our model "over-learned" or memorized our training data, and failed with the new data it hasn't seen before.

__the K parameter we chose as 1 here looks at the closest 1 record and assign the class of that record. This doesn't generalize well to our overall dataset and "overfits" the dataset.

### Where is the validation subset?

In this example, we didn't search for different K values. If we want to optimize the performance of our algorithm, we need to look for the best performing K value.<br/>
__This is done with the validation subset.__ <br/> 
We pick the K value that results in the best validation performance metric of our choice. Then we also look at test set performance as a double check.